// react-dom-server/lib/ReactDOMServer.js
/**
 * Copyright 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @providesModule ReactDOMServer
 */

'use strict';

global.__DEV__ = process.env.NODE_ENV === 'dev';
var ReactServerAsyncRendering = require('./ReactServerAsyncRendering');

var ReactServerNativeComponent = require('./ReactServerNativeComponent');
var ReactDOMServerComponent = require('./ReactDOMServerComponent');
var ReactDOMServerTextComponent = require('./ReactDOMServerTextComponent');

ReactServerNativeComponent.injection.injectGenericComponentClass(ReactDOMServerComponent);
ReactServerNativeComponent.injection.injectTextComponentClass(ReactDOMServerTextComponent);

var EventPluginRegistry = require('./EventPluginRegistry');
var BeforeInputEventPlugin = require('react/lib/BeforeInputEventPlugin');
var ChangeEventPlugin = require('react/lib/ChangeEventPlugin');
var DefaultEventPluginOrder = require('react/lib/DefaultEventPluginOrder');
var EnterLeaveEventPlugin = require('react/lib/EnterLeaveEventPlugin');
var SelectEventPlugin = require('react/lib/SelectEventPlugin');
var SimpleEventPlugin = require('react/lib/SimpleEventPlugin');

/**
 * Inject modules for resolving DOM hierarchy and plugin ordering.
 */
EventPluginRegistry.injectEventPluginOrder(DefaultEventPluginOrder);

/**
 * Some important event plugins included by default (without having to require
 * them).
 */
EventPluginRegistry.injectEventPluginsByName({
  SimpleEventPlugin: SimpleEventPlugin,
  EnterLeaveEventPlugin: EnterLeaveEventPlugin,
  ChangeEventPlugin: ChangeEventPlugin,
  SelectEventPlugin: SelectEventPlugin,
  BeforeInputEventPlugin: BeforeInputEventPlugin
});

var DOMProperty = require('react/lib/DOMProperty');
var SVGDOMPropertyConfig = require('react/lib/SVGDOMPropertyConfig');
var HTMLDOMPropertyConfig = require('react/lib/HTMLDOMPropertyConfig');

DOMProperty.injection.injectDOMPropertyConfig(HTMLDOMPropertyConfig);
DOMProperty.injection.injectDOMPropertyConfig(SVGDOMPropertyConfig);

var ReactDOMServer = {
  createRenderToStringStream: ReactServerAsyncRendering.createRenderToStringStream,
  createRenderToStaticMarkupStream: ReactServerAsyncRendering.createRenderToStaticMarkupStream
};

module.exports = ReactDOMServer;
